modules/selection/src/main/kotlin/com.r3.corda.lib.tokens.selection/memory/services/VaultWatcherService.kt

```diff
@@ -111,36 +125,165 @@ class VaultWatcherService(private val tokenObserver: TokenObserver,
+                    if (largestSeenStateRefIndex != -1) {
+                        syncPoint = SyncPoint(states.last(), specification, toResultSetIndex(states.lastIndex, specification))
+                        if (largestSeenStateRefIndex == states.lastIndex) {
+                            // We have seen the whole list already, so just return last element which we have already seen,
+                            // to simplify processing. Could happen if large number of tokens gets added while we are reading in.
+                            return largestSeenStateRefIndex
+                        }
+                        return largestSeenStateRefIndex + 1 // first unread element
+                    }
+                    LOG.info("Token loading has become out of sync, will re-sync")
+                    return -1
+                }
+                private fun popPriorPageSpecification(): PageSpecification {
+                    pageSpecifications.pollLast()
+                    return if (pageSpecifications.isEmpty()) {
+                        syncPoint = null
+                        pageSpecifications.add(initialSpecification)
+                        initialSpecification
+                    } else {
+                        pageSpecifications.last
+                    }
+                }
+
+                private fun toResultSetIndex(listEntryIndex: Int, specification: PageSpecification): Long {
+                    return specification.pageSize * (specification.pageNumber-1) + listEntryIndex.toLong()
+                }
+
+                private fun toIndexWithinPage(resultSetIndex: Long, specification: PageSpecification): Int {
+                    return (resultSetIndex - ((specification.pageNumber-1) * specification.pageSize)).toInt()
+                }
+
+                fun seachForLargestSeenStateRef(values: List<StateAndRef<FungibleToken>>, element: StateAndRef<FungibleToken>): Int {
+                    var left = 0
+                    var right = values.size - 1
+                    var result = -1
+                    while (left <= right) {
+                        val midPoint: Int = left + (right - left) / 2
+                        val value = values[midPoint]
+                        when (compare(value, element)) {
+                            0  -> return midPoint
+                            -1 -> {
+                                result = midPoint
+                                left = midPoint +1
+                            }
+                            else -> right = midPoint -1
+                        }
+                    }
+                    return result
+                }
+                private fun compare(first: StateAndRef<FungibleToken>, second: StateAndRef<FungibleToken>): Int {
+                    if (first.ref.txhash.bytes.toHexString() < second.ref.txhash.bytes.toHexString()) {
```

Comment:
`SecureHash` is `Comparable`.  So you can use Kotlin `compareValues(first.ref.txhash, second.ref.txhash)`

Reply:
Done. Using compareBy now.